modcc/printer/cprinter.cpp

```diff
@@ -615,6 +616,14 @@ void SimdPrinter::visit(AssignmentExpression* e) {
             out_ << ")";
     }
     else {
+        if (auto rhs = e->rhs()->is_identifier()) {
+            if (auto sym = rhs->symbol()) {
+                if (sym->is_variable() && sym->is_variable()->is_range()) {
+                    out_ << "assign(" << lhs->name() << ", indirect(" << rhs->name() << ", simd_width_))";
+                    return;
+                }
+            }
+        }
         out_ << lhs->name() << " = ";
```

Comment:
Is there ever an instance where we get an external variable on the rhs here?

More generally though. if `assign` were implemented generically in the rhs using `simd_cast` rather than making it specifically only take an indirect expression on the rhs, this clause could be written as
```
if (cast) {
    out_ << assign(" << lhs->name() << ", ";
    e->rhs()->accept(this);
    out_ << assign(" << ")";
}
else ...
```

I noticed when looking at the definition of `assign` in `simd.hpp` that the code currently pulls into the `simd` namespace `simd::detail::simd_impl` and `simd::detail::simd_mask_impl`, which sort of defeats the point of putting them in the `simd::detail` namespace to begin with. Is it possible to rejig that to keep the namespaces clean?

Reply:
Okay. I have extended `assign` to work with scalars and other simd_vectors, which will simplify the generated code even more. However, we can't use the visitor for the `rhs` when it is a range variable. This is because the visitor for range variables automatically casts the indirect expression into a simd_vector because it is necessary when they are arguments to binary expressions.
We also don't expose `sim::detail` with `using` anymore. But I'm not sure that addresses all your concerns on the matter.